### PR TITLE
scribed lecture5

### PR DESCRIPTION
**Pull request by zeshi1990**
Dear instructors,

We have finished scribing for Lecture5. Please review the notes and let us know if any issues need to be fixed. Thanks!

Zeshi, Kun, and Victoria

### CODE REVIEW

**Comment by mrtzh** [comment]
Thanks, can you also add your names to the contributors section in `contributors.tex`?

File: notes/lecture5.tex [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+\]
+How can we project a matrix $A$ onto $B_*$? Formally, we want to solve
+\[
+\min_{X\in B_*}\|A-X\|_F^2
+\]
+Due to the rotational invariance of the Frobenius norm, the solution is obtained by projecting the singular values onto the unit simplex. This operation corresponds to shifting all singular values by the same parameter $\theta$ and clipping values at $0$ so that the sum of the shifted and clipped values is equal to $1$. This algorithm can be found from \cite{Duchi2008}.
+
+\subsubsection{Low-rank matrix completion}
+Suppose we have a partially observable matrix $Y$ and we would like to 
+find its completion form projected in a nuclear norm ball. Formally we have the objective function
```
Comment (mrtzh) [inline notes/lecture5.tex]: Please introduce what `O` is and how it relates to the observed entries. I mainly used the componentwise product with O in the code. For the math, we used $P_{O}$ which is a linear projection onto a subset of the coordinates specified by O.

File: notes/lecture5.tex [hunk 2]
```diff
@@ -0,0 +1,141 @@
+\section{Lecture 5: Conditional Gradient (Frank-Wolfe)}
+\sectionlabel{frankwolfe}
+In this lecture we discussed about the conditional gradient method, also known as the Frank-Wolfe (FW) algorithm. The motivation of using this approach is the projected gradient descent can be computationally inefficient under certain scenarios.
```
Comment (mrtzh) [inline notes/lecture5.tex]: Please add reference to the original paper.

File: notes/lecture5.tex [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+\subsection{Intuition behind Frank-Wolfe algorithm}
+We motivate this lecture by the computational inefficiency that projected gradient can have. With conditional gradient, we are able to sidestep some of these inefficiencies. An intuitive idea of the FW algorithm is as follows. \\
+\\We start from $x_0$. Then, for t = 1 to T steps, we set
```
Comment (mrtzh) [inline notes/lecture5.tex]: Math environments missing around `t=1` and `T`, also please don't use `\\` for new lines.

File: notes/lecture5.tex [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+$$ x_{t+1} = x_t + \eta_t(\bar{x_t}-x_t) $$
+where
+$$ \bar{x_t} = \underset{x \in \domain}{\operatorname{argmin}} f(x_t) + \nabla f(x_t) ^\trans (x-x_t)$$
```
Comment (mrtzh) [inline notes/lecture5.tex]: Please use `\arg\min_{...}`

File: notes/lecture5.tex [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+$$ \bar{x_t} = \underset{x \in \domain}{\operatorname{argmin}} \nabla f(x_t) ^\trans x $$
+Note that we need step size $\eta_t \in [0,1]$ to guarantee $x_{t+1} \in \domain$.
+
+\subsection{Theorem: conditional gradient convergence analysis}
+\begin{theorem}[Convergence Analysis]
+\theoremlabel{convergence}
+Assume:
+\begin{itemize}
+\item $f\colon\domain\to\R$ is convex and $\beta$-smooth
+\item $\exists$ an optimal point, $x^*$, such that $x^* \in\domain$ and $\nabla f(x^*)$ = 0
```
Comment (mrtzh) [inline notes/lecture5.tex]: These assumptions don't read properly. You can't start a sentence with `\exists`. Also don't start sentences with symbols in general.

File: notes/lecture5.tex [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+Using \equationref{convergenceInequal},
+\begin{align*} f(x_{t+1})-f(x) &\leq (1- \frac{2}{t+2}) (f(x_t)-f(x^*))+ \frac{4}{2(t+2)} \beta D^2 \\
+& \leq (1- \frac{2}{t+2}) (\frac{2\beta D^2}{t+2})+ \frac{4}{2(t+2)} \beta D^2 \\
+& = \beta D^2 (\frac{2t}{(t+2)^2} +\frac{2}{(t+2)^2}) \\
+& = 2\beta D^2 \frac{t+1}{(t+2)^2} \\
+& = 2\beta D^2 (\frac{t+1}{t+2}) (\frac{1}{t+2}) \\
+& \leq 2\beta D^2 (\frac{t+2}{t+3}) (\frac{1}{t+2}) \\
+& = 2\beta D^2 \frac{1}{t+3}
+\end{align*}
+Thus, the inequality also holds for the t+1 case.
```
Comment (mrtzh) [inline notes/lecture5.tex]: Again, math environment missing around `t+1`.

File: notes/lecture5.tex [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+\]
+Due to the rotational invariance of the Frobenius norm, the solution is obtained by projecting the singular values onto the unit simplex. This operation corresponds to shifting all singular values by the same parameter $\theta$ and clipping values at $0$ so that the sum of the shifted and clipped values is equal to $1$. This algorithm can be found from \cite{Duchi2008}.
+
+\subsubsection{Low-rank matrix completion}
+Suppose we have a partially observable matrix $Y$ and we would like to 
+find its completion form projected in a nuclear norm ball. Formally we have the objective function
+\[
+\min_{X\in B_*}\frac{1}{2}\|Y-X \odot O\|_F^2
+\]
+And calculate the gradient of this function we will have
```
Comment (mrtzh) [inline notes/lecture5.tex]: Don't start sentences with *and*.

File: notes/lecture5.tex [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+\lemmalabel{lemma}
+\[
+\min_{X\in B_*}\langle\nabla f(X_t)^\trans,X\rangle,\quad B_*=\{X|\|X\|_*\leqslant 1\}
+\]
+The optimal result is \[-\|\nabla f(X_t)\|\] with $X=-u_1v_1^\trans$ being one of the X that minimize the function where
+$u_1$ and $v_1$ are left and right singular vectors corresponding to the max singular value, $\|\nabla f(X_t)\|=\max_{\|X\|_*\leqslant 1}<\nabla f(X_t),X>$.\\
+\end{lemma}
+
+\begin{proof} [Proof of \lemmaref{lemma}.]
+Since the dual norm of a nuclear norm is operator norm,
```
Comment (mrtzh) [inline notes/lecture5.tex]: This is not the proof I gave in the lecture and we haven't introduced dual norms. The proof I gave was based on the fact that the nuclear norm ball is the convex hull of rank-1 matrices. The minimium is attained at a vertex, corresponding to a rank-1 matrix. From there a one line transformation shows that the problem is solved by the top left and right singular vectors of the matrix.

File: notes/lecture5.tex [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+\[
+-\|Y\| = -\nabla f(X_t)=-\sigma_{\max}.
+\]
+Then we show by Singular value decompasition to prove that $X = -u_1v_1^\trans$ is one of the solutions that minimize the function.
+From SVD, $\nabla f(X_t) = U\Sigma V^\trans $, then $\Sigma = U^\trans YV$, we set $X=-u_1v_1^\trans$, thus
+\[
+\langle\nabla f(X_t), X\rangle = \nabla f(X_t)^\trans X = -V\Sigma U^\trans u_1v_1^\trans= -\sigma_1
+\]
+\end{proof}
+According to the lemma, we can see that only the leading singular value and corresponding vectors are used to deprive the argmin X. Thus we can do the rank-1 approximation of this function which makes no difference to final result.\\ \\
```
Comment (mrtzh) [inline notes/lecture5.tex]: deprive the argmin X?

File: notes/lecture5.tex [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+Assume:
+\begin{itemize}
+\item $f\colon\domain\to\R$ is convex and $\beta$-smooth
+\item $\exists$ an optimal point, $x^*$, such that $x^* \in\domain$ and $\nabla f(x^*)$ = 0
+\end{itemize}
+Then, Frank-Wolfe achieves
+$$ f(x_t) - f(x^*) \leq \frac{2\beta D^2}{t+2}$$
+with step size 
+$$\eta_t = \frac{2}{t+2}$$
+where D is the diameter of $\domain$, defined:
```
Comment (mrtzh) [inline notes/lecture5.tex]: Math environment missing around `D`.

File: notes/lecture5.tex [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+$$ D = \max_{x-y \in \domain} \| x-y \|$$
+\end{theorem}
+Note that we can trade our assumption of the existence of $x^*$ for a dependence on L in our bound.
```
Comment (mrtzh) [inline notes/lecture5.tex]: Math environment missing around `L`.

File: notes/lecture5.tex [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+We now recall the definition of $D$ from \theoremref{convergence} and observe that $\| \bar{x_t} - x_t \| ^2 \leq D^2$. Thus, we rewrite the inequality:
+$$ f(x_{t+1}) \leq f(x_t) + \eta_t \nabla f(x_t) ^\trans (x_t^* - x_t) + \frac{\eta_t^2\beta D ^2}{2} $$
+Because of convexity, we also have that
+$$ \nabla f(x_t) ^\trans (x^*-x_t) \leq f(x^*) - f(x_t) $$
+Thus,
+\begin{equation}
+\equationlabel{convergenceInequal}
+f(x_{t+1})-f(x^*) \leq (1-\eta_t)(f(x_t)-f(x^*)) + \frac{\eta_t^2\beta D ^2}{2}
+\end{equation}
+\\We use induction in order to prove $f(x_t)-f(x^*) \leq \frac{2\beta D^2}{t+2}$  based on \equationref{convergenceInequal} above.\\
```
Comment (mrtzh) [inline notes/lecture5.tex]: Please don't start a line with \\.

File: notes/lecture5.tex [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+& = 2\beta D^2 (\frac{t+1}{t+2}) (\frac{1}{t+2}) \\
+& \leq 2\beta D^2 (\frac{t+2}{t+3}) (\frac{1}{t+2}) \\
+& = 2\beta D^2 \frac{1}{t+3}
+\end{align*}
+Thus, the inequality also holds for the t+1 case.
+
+\end{proof}
+
+\subsection{Examples}
+The code for the following examples can be found \href{https://ee227c.github.io/code/lecture4.html}{here}.
```
Comment (mrtzh) [inline notes/lecture5.tex]: I believe this is the wrong URL. It should be lecture5.html.

File: notes/lecture5.tex [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+The norm can be computed from the singular value decomposition of $A$.
+We denote the unit ball of the nuclear norm by 
+\[
+B_*^{m\times n}=\{A\in\mathbb{R}^{m\times n} \mid \|A\|_*\le 1\}.
+\]
+How can we project a matrix $A$ onto $B_*$? Formally, we want to solve
+\[
+\min_{X\in B_*}\|A-X\|_F^2
+\]
+Due to the rotational invariance of the Frobenius norm, the solution is obtained by projecting the singular values onto the unit simplex. This operation corresponds to shifting all singular values by the same parameter $\theta$ and clipping values at $0$ so that the sum of the shifted and clipped values is equal to $1$. This algorithm can be found from \cite{Duchi2008}.
```
Comment (mrtzh) [inline notes/lecture5.tex]: can be found *in*

File: notes/lecture5.tex (hunk 1 above)
Comment (mrtzh) [reply notes/lecture5.tex]: Generally, this is not an adequate introduction of the problem. Please give sufficient description for the reader to be able to follow.

File: notes/lecture5.tex [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+-\|Y\| = -\nabla f(X_t)=-\sigma_{\max}.
+\]
+Then we show by Singular value decompasition to prove that $X = -u_1v_1^\trans$ is one of the solutions that minimize the function.
+From SVD, $\nabla f(X_t) = U\Sigma V^\trans $, then $\Sigma = U^\trans YV$, we set $X=-u_1v_1^\trans$, thus
+\[
+\langle\nabla f(X_t), X\rangle = \nabla f(X_t)^\trans X = -V\Sigma U^\trans u_1v_1^\trans= -\sigma_1
+\]
+\end{proof}
+According to the lemma, we can see that only the leading singular value and corresponding vectors are used to deprive the argmin X. Thus we can do the rank-1 approximation of this function which makes no difference to final result.\\ \\
+Power method works well in approximate the leading singular value and corresponding left and right vetors.\\
```
Comment (mrtzh) [inline notes/lecture5.tex]: Grammar

Comment (mrtzh) [reply notes/lecture5.tex]: Also don't use `\\`, just use an empty line.

File: notes/lecture5.tex [hunk 16]
```diff
@@ -0,0 +1,141 @@
...
+Here is the introduction of the process of SVD power method.
```
Comment (mrtzh) [inline notes/lecture5.tex]: SVD and power method are two different things. You want to describe the power method, not SVD.

File: notes/lecture5.tex [hunk 17]
```diff
@@ -0,0 +1,141 @@
...
+\begin{itemize}
+\item $w_k=Av_{k-1}, \alpha_k=\|w_k\|, u_k={\alpha_k}^{-1}w_k$
+\item $z_k=A^\trans, \beta_k = \|z_k\|, v_k=\beta_kz_k$
+\end{itemize}
+$k=1, 2, 3, 4, 5,... , \sigma_{max}=\beta_k, v_k=\delta_{2k}\sum\limits_{i=1}^r \sigma_j ^{2k}y_jv_j, u_k=\delta_{2k+1}\sum\limits_{i=1}^r \sigma_j ^{2k+1}y_jv_j$\\
```
Comment (mrtzh) [inline notes/lecture5.tex]: I don't know what you're describing here, but it's not what we did in class. Please look at the code for a proper description of the algorithm.

File: notes/notes.bib [hunk 18]
```diff
@@ -7,3 +7,29 @@ @article{bubeck
...
+doi = {10.1145/1390156.1390191},
+isbn = {9781605582054},
+pages = {272--279},
+publisher = {ACM Press},
+title = {{Efficient projections onto the l 1 -ball for learning in high dimensions}},
+url = {http://portal.acm.org/citation.cfm?doid=1390156.1390191},
+year = {2008}
+}
+
+@article{Bentbib2015,
```
Comment (mrtzh) [inline notes/notes.bib]: This is not the right reference. It's probably best to leave it out. I may add a more appropriate reference later.

File: notes/lecture5.tex (hunk 8 above)
Comment (zeshi1990) [reply notes/lecture5.tex]: Our original version was based on the nuclear norm ball is the convex hull of rank-1 matrices. But we do not have a solid proof of this fact. Do you have any suggestions on references about this proof?

Comment (mrtzh) [reply notes/lecture5.tex]: It follows from the singular value decomposition. You can state it as a fact without proof.

**Comment by zeshi1990** [comment]
Dear instructor,

We have revised the note for Lecture 5. Please let us know if any issues arise. Thanks!

-Zeshi, Kun, Victoria

**Comment by mrtzh** [comment]
Thanks for your work on this!

**Pull request merged**